src/System.Net.Security/src/System/Net/Security/SslStreamPal.Unix.cs

```diff
@@ -160,6 +160,16 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
                     done = Interop.OpenSsl.DoSslHandshake(((SafeDeleteSslContext)context).SslContext, inputBuffer.token, inputBuffer.offset, inputBuffer.size, out output, out outputSize);
                 }
 
+                // When the handshake is done, and the context is server, check if the alpnHandle target was set to null during ALPN.
+                // If it was, then that indiciates ALPN failed, send failure.
+                // We have this workaround, as openssl supports terminating handshake only from version 1.1.0,
+                // whereas ALPN is supported from version 1.0.2.
+                SafeSslHandle sslContext = ((SafeDeleteSslContext)context).SslContext;
+                if (done && sslAuthenticationOptions.IsServer && sslAuthenticationOptions.ApplicationProtocols != null && sslContext.AlpnHandle.IsAllocated && sslContext.AlpnHandle.Target == null)
+                {
+                    throw Interop.OpenSsl.CreateSslException(SR.net_alpn_failed);
```

Comment:
I still don't understand why we're throwing rather than returning here.  It's incurring the cost of throwing, but it's just going to be caught a few lines down and turned into something that's then returned.  Why not instead just return it:
```C#
return new SecurityStatusPal(
    SecurityStatusPalErrorCode.InternalError, // or whatever error code makes sense here
    Interop.OpenSsl.CreateSslException(SR.net_alpn_failed));
```
?

Reply:
changed.